lint/location.go

```diff
@@ -14,43 +14,97 @@
 
 package lint
 
-import "fmt"
+import (
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
 
-// Location describes a location in a source code file.
+// SyntaxLocation returns the location of the syntax definition in a file
+// descriptor.
 //
-// Note: positions are one-based.
-type Location struct {
-	Start Position `json:"start_position" yaml:"start_position"`
-	End   Position `json:"end_position" yaml:"end_position"`
+// If the location can not be found (for example, because there is no syntax
```

Comment:
Why not `nil`? Especially that we are returning a pointer.

Reply:
Because this is about usefulness. The idea is that I _always_ want a valid location of some kind.

If someone sends a `Problem` and there is some unforeseen issue, the graceful fallback behavior is "put the error at the top of the file", as opposed to "return nil and probably panic later".

Reply:
But in the `Problem`, we already do the fallback to get the location from the `Descriptor` if `Location == nil`, which in this case is the `FileDescriptor`.

Reply:
Right, and this is what we _use_ for that. It is therefore important that we return a valid location.

Reply:
Agree that we should return nil. We can still provide convenience in the core linter (for example, as we already do: if locaiton is nil, we set it to the location of the descriptor).

If you truly want to have this fallback to the start location of the file, you can name the function something like `SyntaxLocationOrFileStart`, but I don't see the advantage of doing this.

The user will be using this to attach locations to problems. If we gracefully handle the `nil` case in the core linter code, we shouldn't have to worry about panics.

In what case do you see returning the start location as useful?

Reply:
Convinced, will return `nil`. Waiting for #194 to be merged before making changes here.